Zend/zend_compile.c

```diff
@@ -5372,21 +5372,41 @@ void zend_compile_match(znode *result, zend_ast *ast) /* {{{ */
 			}
 		}
 
+		zend_ast *arm_body_statements;
+		zend_ast *arm_body_expr_ast;
+		if (arm_body_ast->kind == ZEND_AST_MATCH_BLOCK) {
+			arm_body_statements = arm_body_ast->child[0];
+			arm_body_expr_ast = arm_body_ast->child[1];
```

Comment:
This currently treats match within a match as a fatal error and gives a misleading error message, as a consequence of the optional semicolon.

Aside: the error is for the match starting on line 6 and ending on line 11, but the error message had line 9. 

```php
<?php
function test ($x, $y) {
    $result = match ($x) {
        1 => {
            echo "x is 1\n";
            match($y) {
                2 => {
                    $z = $y + 1;
                    $x * $z  // Observed: Fatal error: Match block must return a value. Did you mean to omit the last semicolon?
                }
            }
        }
    };
}
```

Reply:
Yeah I know: https://github.com/php/php-src/pull/5407#issuecomment-616612763

One improvement is to only suggest removing the semicolon when the last `arm_body_statements` is an expression. Unfortunately it's not easy to find out if this is the case.

> Aside: the error is for the match starting on line 6 and ending on line 11, but the error message had line 9.

Oh, I'll take a look at that.

Reply:
We'll discuss that in another RFC in detail 🙂